### PR TITLE
Helm HA Install

### PR DESCRIPTION
**Pull request by jgreat**
Don't make this live until Rancher 2.0.5 is live and the Helm catalog (rancher/server-charts) has been updated. 

Installation guide for HA with Helm.

* node requirements
* rke (keeping it very simple by default)
* helm init
* helm install rancher-stable/rancher

* Air Gap/Private Registry steps, when required.
* Basic Troubleshooting guide to when the chart doesn't work.

### CODE REVIEW

**Comment by JasonvanBrackel** [comment]
@jgreat Do you have time to resolve these conflicts this week?

File: content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+title: High Availability Installation
+weight: 275
+draft: true
+---
+
+When installed as a Deployment in a Kubernetes cluster, Rancher will take integrate with the cluster's etcd database and Kubernetes scheduling for High-Availability.
+
+This procedure walks you through setting up a 3-node cluster with RKE and installing the Rancher chart the Helm package manager.
+
+> NOTE: For the best performance, we recommend this Kubernetes cluster be dedicated only the Rancher workload.
```
Comment (superseb) [inline content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md]: typo/sentence

File: content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+## Recommended Architecture
+
+* DNS for Rancher should resolve to a Layer 4 Load Balancer
+* The Load Balancer should forward ports 80 and 443 TCP to all 3 nodes in the Kubernetes cluster.
+* The ingress controller will redirect http port 80 to https and terminate SSL/TLS on port 443.
```
Comment (superseb) [inline content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md]: Ingress is a term and should be written as Ingress

File: content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+* The ingress controller will forward traffic to port 80 on the Pod in the Rancher Deployment.
+
+![Rancher HA]({{< baseurl >}}/img/rancher/ha/rancher2ha.svg)
+
+## Required Tools
+
+The following CLI tools are required for this install. Please make sure these tools installed and available in your `$PATH`
```
Comment (superseb) [inline content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md]: tools vs binaries, naming is using binary names not tools names.

File: content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+### Load Balancer
+
+RKE will configure an ingress-controller pod, on each of your nodes. The ingress-controller pods are bound to ports 80 and 443 tcp on the host network and are the entry point for https traffic to the Rancher server.
+
+Configure a load balancer as a basic Layer 4 tcp forwarder. The exact configuration will vary depending on your environment.
+
+#### Examples
+
+* [Amazon NLB](nlb/)
```
Comment (superseb) [inline content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md]: If you are putting in examples, also put in the original example of NGINX Layer 4

File: content/rancher/v2.x/en/installation/_index.md [hunk 5]
```diff
@@ -4,38 +4,22 @@ weight: 50
...
+### Installation Options
 
-- [Single Node Installation]({{< baseurl >}}/rancher/v2.x/en/installation/single-node-install)
+- [Single Node Installation](single-node-install/)
 
-	In this simple install scenario, you install Rancher on a single Linux host.
+	Install Rancher on a single Linux host.
 
-- [Single Node Installation with External Load Balancer]({{< baseurl >}}/rancher/v2.x/en/installation/single-node-install-external-lb)
+-  [High Availability Installation](ha-server-install-helm/)
```
[the remark this answers is not in the document]
Comment (jgreat) [reply content/rancher/v2.x/en/installation/_index.md]: That install option is listed under the High Availability Installation subsection.

File: content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md (hunk 3 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md]: I disagree with this one.  Tools and binaries are pretty much synonymous in this particular case.  You may install some off these as standalone binaries, but you can also use package managers.

File: content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md (hunk 4 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md]: There is no nginx example for l4 load balancers in our current docs.

File: content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md (hunk 1 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md]: Good catch. Fixing it.

File: content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md (hunk 2 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha-server-install-helm/_index.md]: fixed.

File: content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+In addition to the ports listed above these ports must be open between nodes.
+
+| Protocol | Ports | Description |
+| --- | --- | --- |
+| tcp | 2379-2380 | etcd |
+| udp | 8472 | overlay networking |
+| tcp | 10250 | kubelet |
+<br/>
+
```
[the remark this answers is not in the document]
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md]: This is updated. Created a new requirements file.

File: content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md (hunk 4 above)
Comment (superseb) [reply content/rancher/v2.x/en/installation/ha-server-install-helm/create-nodes-and-load-balancer/_index.md]: https://rancher.com/docs/rancher/v2.x/en/installation/ha-server-install/ is using it

File: content/rancher/v2.x/en/installation/ha/_index.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
+---
+title: High Availability Installation
+weight: 275
+draft: true
+---
+
+When installed as a Deployment in a Kubernetes cluster, Rancher will take integrate with the cluster's etcd database and Kubernetes scheduling for High-Availability.
```
Comment (deniseschannon) [inline content/rancher/v2.x/en/installation/ha/_index.md]: ` Rancher will take integrate with` - not sure what this means

File: content/rancher/v2.x/en/installation/ha/_index.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+This procedure walks you through setting up a 3-node cluster with RKE and installing the Rancher chart with the Helm package manager.
+
+> NOTE: For the best performance, we recommend this Kubernetes cluster is dedicated to only the Rancher workload.
```
Comment (deniseschannon) [inline content/rancher/v2.x/en/installation/ha/_index.md]: for notes, the format is usually 
```
> **Note:**
```

this sentence seems awkward: `we recommend this Kubernetes cluster is dedicated to only the Rancher workload.`

maybe `we recommend this Kubernetes cluster to be dedicated to only the Rancher workload.`

File: content/rancher/v2.x/en/installation/ha/_index.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+* DNS for Rancher should resolve to a Layer 4 Load Balancer
+* The Load Balancer should forward ports 80 and 443 TCP to all 3 nodes in the Kubernetes cluster.
+* The Ingress controller will redirect http port 80 to https and terminate SSL/TLS on port 443.
+* The Ingress controller will forward traffic to port 80 on the Pod in the Rancher Deployment.
+
+![Rancher HA]({{< baseurl >}}/img/rancher/ha/rancher2ha.svg)
+
+## Required Tools
+
+The following CLI tools are required for this install. Please make sure these tools installed and available in your `$PATH`
```
Comment (deniseschannon) [inline content/rancher/v2.x/en/installation/ha/_index.md]: `Please make sure these tools installed` -> `Please make sure these tools are installed`

File: content/rancher/v2.x/en/installation/ha/_index.md (hunk 7 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha/_index.md]: how's this

When installed in a Kubernetes cluster, Rancher will integrate with the cluster's etcd database and take advantage of Kubernetes scheduling for High-Availability.

File: content/rancher/v2.x/en/installation/ha/_index.md (hunk 8 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha/_index.md]: > **Note:** For the best performance, we recommend this Kubernetes cluster be dedicated only to the Rancher workload.

File: content/rancher/v2.x/en/installation/ha/_index.md (hunk 9 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha/_index.md]: fixed

File: content/rancher/v2.x/en/installation/ha/helm-rancher/_index.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+#### Certificates from Files (Kubernetes Secret)
+
+Create Kubernetes Secrets from your own certificates for Rancher to use.
+
+> NOTE: The common name for the cert will need to match the `hostname` option or the ingress controller will fail to provision the site for Rancher.
+
+Set `hostname` and `ingress.tls.source=secret`
+
+> NOTE: If you are using a Private CA signed cert, add `--set privateCA=true`
```
Comment (Oats87) [inline content/rancher/v2.x/en/installation/ha/helm-rancher/_index.md]: This is kind of hidden, I would maybe have two of the `helm install` examples, one with and one without this.

File: content/rancher/v2.x/en/installation/ha/helm-rancher/tls-secrets/_index.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+kubectl -n cattle-system create secret tls tls-rancher-ingress \
+--cert=./tls.crt \
+--key=./tls.key
+```
+
+### Private CA Signed - Additional Steps
+
+If you are using a private CA, Rancher will need to have a copy of the CA cert to include when generating agent configs.
+
+Copy the CA cert into a file named `cacerts.pem` and use `kubectl` to create the `tls-ca` secret in the `cattle-system` namespace.
```
Comment (Oats87) [inline content/rancher/v2.x/en/installation/ha/helm-rancher/tls-secrets/_index.md]: Is this too confusing for the end user? Having to copy their "ca" to "cacerts.pem", perhaps we modify the create secret to `--from-file=cacerts.pem=<your-ca-file>` or something?

File: content/rancher/v2.x/en/installation/ha/helm-rancher/_index.md (hunk 10 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha/helm-rancher/_index.md]: I don't know if this is hidden, it is in a highlighted block.  Lets go with what we have now, and if there is confusion we can figure out a way to make it more clear.

File: content/rancher/v2.x/en/installation/ha/helm-rancher/tls-secrets/_index.md (hunk 11 above)
Comment (jgreat) [reply content/rancher/v2.x/en/installation/ha/helm-rancher/tls-secrets/_index.md]: Can we capture this as an improvement for a later revision. This PR is pretty complex. I don't want to make anymore changes in this PR unless something just doesn't work.

File: content/rancher/v2.x/en/installation/ha/helm-rancher/_index.md (hunk 10 above)
Comment (Oats87) [reply content/rancher/v2.x/en/installation/ha/helm-rancher/_index.md]: Okay, works for me. Reason why I point it out is I missed it the first time around with my custom-CA

File: content/rancher/v2.x/en/installation/ha/helm-rancher/tls-secrets/_index.md (hunk 11 above)
Comment (Oats87) [reply content/rancher/v2.x/en/installation/ha/helm-rancher/tls-secrets/_index.md]: Yes, just trying to ensure that the end-user experience is as error-free (i.e. if they don't look at every step and accidentally skip something) as possible. Looks good to me, all 3 methods work when the directions are followed verbatim

**Pull request merged**